Create inline-arrays.md (#7064)
dotnet/csharplang: proposals/safe-fixed-size-buffers.md

specific element with appropriate "readonly-ness", which prevents direct assignment to the elements when language rules don't permit that.

The resulting span instance will have a length equal to the size declared on the fixed-size buffer.
Indexing into the span with a constant expression outside of the declared fixed-size buffer bounds is a compile time error.

The *safe-to-escape* scope of the value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.

### Fixed-size buffers in expressions

Member lookup of a fixed-size buffer member proceeds exactly like member lookup of a field.

A fixed-size buffer can be referenced in an expression using a *simple_name* or a *member_access* .

When an instance fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `this.I`, where `I` is the fixed-size buffer member. When a static fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `E.I`, where `I` is the fixed-size buffer member and `E` is the declaring type.

#### Non-readonly fixed-size buffers

> In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies an instance fixed-size member,

Reviewer: ```suggestion
In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies an instance non-readonly fixed-size member,
```

I know the section title says *non-readonly*, but I found this confusing. This sentence by itself makes it sound like the rule applies even if the fixed-size member is readonly.

then `E.I` is evaluated and classified as follows:

- If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.
  The value can be used to access members’ elements.
- Otherwise, `E` is classified as a variable and the result of the expression is classified as a value of type `System.Span<S>`, where S is the element type of `I`.
  The value can be used to access members’ elements.

In a member access of the form `E.I`, if `E` is of a class type and a member lookup of `I` in that class type identifies an instance fixed-size member,
then `E.I` is evaluated and classified as a value of type `System.Span<S>`, where S is the element type of `I`.

In a member access of the form `E.I`, if member lookup of `I` identifies a static fixed-size member,
then `E.I` is evaluated and classified as a value of type `System.Span<S>`, where S is the element type of `I`.

#### Readonly fixed-size buffers

> When a *field_declaration* includes a `readonly` modifier, the member introduced by the fixed_size_buffer_declarator is a ***readony fixed-size buffer***.

Reviewer: ```suggestion
When a *field_declaration* includes a `readonly` modifier, the member introduced by the fixed_size_buffer_declarator is a ***readonly fixed-size buffer***.
```

A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.

Depending on the situation (details are specified below), an access to a fixed-size buffer member is classified as a value (never a variable) of either
`System.ReadOnlySpan<S>` or `System.Span<S>`, where S is the element type of the fixed-size buffer. Both types provide indexers returning a reference to a
specific element with appropriate "readonly-ness", which prevents direct assignment to the elements when language rules don't permit that.

This limits the set of types that can be used as a fixed-size buffer element type to types that can be used as type arguments. For example, a pointer type cannot be used as an element type.

The resulting span instance will have a length equal to the size declared on the fixed-size buffer.
Indexing into the span with a constant expression outside of the declared fixed-size buffer bounds is a compile time error.

The *safe-to-escape* scope of the value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.

> ### Fixed-size buffers in expressions

Reviewer 2: Do we need to spec the behavior of `nameof(C.Field)` where `Field` is a fixed sized buffer? Or does that just fall out from the member access rules?

Member lookup of a fixed-size buffer member proceeds exactly like member lookup of a field.

A fixed-size buffer can be referenced in an expression using a *simple_name* or a *member_access* .

When an instance fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `this.I`, where `I` is the fixed-size buffer member. When a static fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `E.I`, where `I` is the fixed-size buffer member and `E` is the declaring type.

#### Non-readonly fixed-size buffers

In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies a non-readonly instance fixed-size member,
then `E.I` is evaluated and classified as follows:

- If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.

>   The value can be used to access members’ elements.

Reviewer 2: The value can be used to access memebr's elements but not must be. For example my reading believes the following should work: 

```csharp
S local = default;
Span<int> span = s.Field;

struct S {
  public int[5] Field;
}

```

Is that correct?

Author: > The value can be used to access memebr's elements but not must be. For example my reading believes the following should work:

Yes, ```Span<int> span = s.Field``` will work.

``` C#
public partial class C
{
    public int buffer1[10];
    public readonly int buffer2[10];
}
```
will be emitted as  fields of a specially decorated struct type.

Equivalent C# code will be:

``` C#
public partial class C
{
    public Buffer10<int> buffer1;
    public readonly Buffer10<int> buffer2;
}

[System.Runtime.CompilerServices.InlineArray(10)]
public struct Buffer10<T>
{
    private T _element0;

    [UnscopedRef]
    public System.Span<T> AsSpan()
    {
        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
    }

    [UnscopedRef]
    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
    {
        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
    }
}
```

The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types

> will be "speakable" in order to support consumption from other languages. 

Reviewer 2: Agree they should be speakable. I was thinking about what the best namespace would be for these types when the compiler generates them. Ended up thinking that `System.Runtime.CompilerServices` is a good candidate. Mostly cause that is the likely place the libraries team would want to put the predefined types. That would mean 

1. The libraries team could use the compiler to effectively generate the types they need vs. hand coding them since we're putting them in the right place.
2. It's a consistent location for the types to always exist in

My other thought was `Microsoft.CodeAnalysis`.

Reviewer 3: > Mostly cause that is the likely place the libraries team would want to put the predefined types.

From the BCL point of view, I do not think we want the compiler to ever auto-generate types in the `System.Runtime.CompilerServices` namespace. That's would be a factory for type-name collisions.

Can we make the physicial type of the field a hidden internal implementation detail and only expose these fields publicly via the Span and ReadOnlySpan accessors?

Reviewer 4: > Can we make the physicial type of the field a hidden internal implementation detail and only expose these fields publicly via the Span and ReadOnlySpan accessors?

This is my preference as well.  Basically mirror what we do today for unsafe fixed buffers: a developer writes `fixed int buffer[10]`, the compiler generates an unspeakable type for that, and the effective type visible to consuming code is `int*`.  For safe buffers, it could be the same approach: the developer writes `int buffer[10]`, the compiler generates an unspeakable type for that, and the effective type visible to consuming code is either `Span<int>` or `ReadOnlySpan<int>` depending on whether the field is readonly.

> The libraries team could use the compiler to effectively generate the types they need vs. hand coding them since we're putting them in the right place.

Handcoding these is trivial. _If_ we wanted to expose any publicly for size/sharing reasons, it would be easy to do so without the compiler feature.  As noted earlier, I'd also be concerned about using the compiler feature to implement these, as it would either be limiting for what the compiler could emit on them in the future (any additions the compiler made here would implicitly impact core library public surface area), or if we didn't want to constrain the compiler in that fashion (I don't want to), we wouldn't use it for this, anyway.

## Motivation

This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.

## Detailed Design (Option 1)

The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
will be extended to allow specifying the size of the buffer:

``` antlr
field_declaration
    : attributes? field_modifier* type variable_declarators ';'
    ;

field_modifier
    : 'new'
    | 'public'
    | 'protected'
    | 'internal'
    | 'private'
    | 'static'
    | 'readonly'
    | 'volatile'
    | unsafe_modifier   // unsafe code support
    ;

variable_declarators
    : variable_declarator (',' variable_declarator)*
    ;
    
variable_declarator
    : identifier ('=' variable_initializer)?
    | fixed_size_buffer_declarator
    ;
    

> fixed_size_buffer_declarator

Author: > Was the sharing intentional?

Yes.

The elements of a fixed-size buffer shall be laid out sequentially in memory as though they are elements of an array.

A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.

Depending on the situation (details are specified below), an access to a fixed-size buffer member is classified as a value (never a variable) of either
`System.ReadOnlySpan<S>` or `System.Span<S>`, where S is the element type of the fixed-size buffer. Both types provide indexers returning a reference to a
specific element with appropriate "readonly-ness", which prevents direct assignment to the elements when language rules don't permit that.

This limits the set of types that can be used as a fixed-size buffer element type to types that can be used as type arguments. For example, a pointer type cannot be used as an element type.

> The resulting span instance will have a length equal to the size declared on the fixed-size buffer.

Reviewer 5: > The resulting span instance will have a length equal to the size declared on the fixed-size buffer.

It seems this might not be always true if the Span is created by user-supplied buffer type (from metadata).

Author: > It seems this might not be always true if the Span is created by user-supplied buffer type (from metadata).

I am not sure why that would make the difference. The attribute emitted on the type is supposed to match the size. Malformed metadata isn't something that the language spec is concerned about.

Indexing into the span with a constant expression outside of the declared fixed-size buffer bounds is a compile time error.

The *safe-to-escape* scope of the value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.

### Fixed-size buffers in expressions

Member lookup of a fixed-size buffer member proceeds exactly like member lookup of a field.

A fixed-size buffer can be referenced in an expression using a *simple_name* or a *member_access* .

When an instance fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `this.I`, where `I` is the fixed-size buffer member. When a static fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `E.I`, where `I` is the fixed-size buffer member and `E` is the declaring type.

#### Non-readonly fixed-size buffers

> In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies a non-readonly instance fixed-size member,

Author: > The terms "fixed-size buffer", "fixed-size member" all seem ambiguous with the existing feature [fixed-size buffers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers). The full unambiguous name is "safe fixed-size buffers", but that's long, should we use a different shorter name?

I will stick to the current terminology, but will make a note at the top to set the context

then `E.I` is evaluated and classified as follows:

> - If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.
>   The value can be used to access members’ elements.
> - Otherwise, `E` is classified as a variable and the result of the expression is classified as a value of type `System.Span<S>`, where S is the element type of `I`.
>   The value can be used to access members’ elements.

Reviewer 5: ```suggestion
- If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.
  The value can be used to access the member's elements.
- Otherwise, `E` is classified as a variable and the result of the expression is classified as a value of type `System.Span<S>`, where S is the element type of `I`.
  The value can be used to access the member's elements.
```

will be "speakable" in order to support consumption from other languages. 

A code generated for an access like: 
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer1[1] = val;
    }

    int M2()
    {
        return buffer2[1];
    }
}
```

will be equivalent to:
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer.AsSpan()[1] = val;
    }

    int M2()
    {
        return buffer2.AsReadOnlySpan()[1];
    }
}
```

#### Metadata import

When compiler imports a field declaration of type *T* and the following conditions are all met:
- *T* is a struct type decorated with the ```InlineArray``` attribute, and
- The first instance field declared within *T* has type *F*, and
- There is a ```public System.Span<F> AsSpan()``` within *T*, and 
- There is a ```public readonly System.ReadOnlySpan<T> AsReadOnlySpan()``` or ```public System.ReadOnlySpan<T> AsReadOnlySpan()``` within *T*. 

the field will be treated as C# fixed-size buffer with element type *F*. Otherwise, the field will be treated as a regular field of type *T*.

> ## Detailed Design (Option 2)

Reviewer 6: >(Option 2) [](http://example.com/codeflow?start=19&length=10)

Are the differences between the two options limited to element access? If so, consider removing the common items from the second option, so the differences are clear.

Author: > Are the differences between the two options limited to element access?

No they are not limited to just to element access. The first option doesn't have special treatment of fixed-size buffer types, there is a difference in field type from the language point of view, any occurrences of fixed-size buffer types are  affected (not just fields), etc.

    }

    [UnscopedRef]
    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
    {
        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
    }
}
```

This limits the set of types that can be used as a fixed-size buffer element type to types that can be used as type arguments. For example, a pointer type cannot be used as an element type.

### Element access

The [Element access](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#11710-element-access) will be extended
to support fixed-size buffer element access.

An *element_access* consists of a *primary_no_array_creation_expression*, followed by a “`[`” token, followed by an *argument_list*, followed by a “`]`” token. The *argument_list* consists of one or more *argument*s, separated by commas.

```ANTLR
element_access
    : primary_no_array_creation_expression '[' argument_list ']'
    ;
```

The *argument_list* of an *element_access* is not allowed to contain `ref` or `out` arguments.

> An *element_access* is dynamically bound ([§11.3.3](expressions.md#1133-dynamic-binding)) if at least one of the following holds:

Author: This is a direct quote. It is not my goal to maintain links

#### Fixed-size buffer element access

For a fixed-size buffer element access, the *primary_no_array_creation_expression* of the *element_access* must be a variable or value of a fixed-size buffer type. Furthermore, the *argument_list* of a fixed-size buffer element access is not allowed to contain named arguments. The *argument_list* must contain a single expression, and the expression must be 
- either of type `int`, or
- must be implicitly convertible to `int`, or
- must be of type ```System.Index```, or 

> - must be of type ```System.Range```. 

Reviewer 6: Consider dropping the "either" and "must be" from the bullets.
> - of type `int`, or
> - implicitly convertible to `int`, or
> - of type `System.Index`, or
> - of type `System.Range`.

> - either of type `int`, or
> - must be implicitly convertible to `int`, or
> - must be of type ```System.Index```, or 
> - must be of type ```System.Range```. 

Reviewer 6: Consider dropping the "either" and "must be" from the bullets.
```suggestion
- of type `int`, or
- implicitly convertible to `int`, or
- of type ```System.Index```, or 
- of type ```System.Range```. 
```

If *primary_no_array_creation_expression* is a variable, result of evaluating a fixed-size buffer element access is result of indexing into a ```System.Span<T>``` returned by *AsSpan* method with the same argument. 

> If *primary_no_array_creation_expression* is a value, result of evaluating a fixed-size buffer element access is result of indexing into a ```System.ReadOnlySpan<T>``` returned by *AsReadOnlySpan* with the same argument. 

Reviewer 6: Perhaps "_the_ result of evaluating a fixed-size buffer element access is _the_ result of ..." for the two cases above.

> If *primary_no_array_creation_expression* is a variable, result of evaluating a fixed-size buffer element access is result of indexing into a ```System.Span<T>``` returned by *AsSpan* method with the same argument. 
> 
> If *primary_no_array_creation_expression* is a value, result of evaluating a fixed-size buffer element access is result of indexing into a ```System.ReadOnlySpan<T>``` returned by *AsReadOnlySpan* with the same argument. 

Reviewer 6: ```suggestion
If *primary_no_array_creation_expression* is a variable, the result of evaluating a fixed-size buffer element access is the result of indexing into a ```System.Span<T>``` returned by *AsSpan* method with the same argument. 

If *primary_no_array_creation_expression* is a value, the result of evaluating a fixed-size buffer element access is the result of indexing into a ```System.ReadOnlySpan<T>``` returned by *AsReadOnlySpan* with the same argument. 
```

just as it would if the backing data was accessed as a field.

### Conversions

A new conversion from expression will be added.

There is an implicit conversion from expression representing a variable of a fixed-size buffer type to the following types:
- ```System.Span<T>``` - the result of conversion is a span returned by *AsSpan* method.
- ```System.ReadOnlySpan<T>``` - the result of conversion is a span returned by *AsReadOnlySpan* method.

There is an implicit conversion from expression representing a value of a fixed-size buffer type to ```System.ReadOnlySpan<T>``` type.
The result of conversion is a span returned by *AsReadOnlySpan* method.

### List patterns

[List patterns](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md) will not be supported for instances
of fixed-size buffer types. In order to be able to take advantage of list patterns on fixed-size buffer types, a target expression must be
explicitly converted either to ```System.Span<T>```, or to ```System.ReadOnlySpan<T>``` type first. The reason is the fact that pattern evaluation 

> liberally makes copies of the target expression, and it would be impossble to slice the copy into ```System.Span<T>```, but intent of the user might

Reviewer 6: ```suggestion
liberally makes copies of the target expression, and it would be impossible to slice the copy into ```System.Span<T>```, but intent of the user might
```

> liberally makes copies of the target expression, and it would be impossble to slice the copy into ```System.Span<T>```, but intent of the user might

Reviewer 6: ```suggestion
liberally makes copies of the target expression, and it would be impossible to slice the copy into ```System.Span<T>```, but intent of the user might
```

(a statement or an expression) and they are often reused. When a target expression is copied is unspecified, and, often, depends on complexity of
the pattern and the evaluation strategy chosen by compiler. 

### MemoryMarshal.CreateReadOnlySpan

We might want to specially recognize ```System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan``` method and allow
passing a reference to a readonly location as its first argument. This should allow users to define their own fixed-size buffer
types. For example, in order to be able to share a type across multiple assemblies and also include custom APIs into its definition.
This might also help framework to define our shared fixed-size buffer types.

### Definite assignment checking

Regular definite assignment rules are applicable to variables that have a fixed-size buffer type. 

## Open design questions

### Initializer

Should we support initialization at declaration site with, perhaps, [collection literals](https://github.com/dotnet/csharplang/blob/main/proposals/collection-literals.md)?

## Alternatives

> ### Method or property group like approach in the language

Reviewer 5: This alternative looks like Option 2, what's the difference?

Author: Starting with

> Safe fixed-size buffer accesses have their own classification (just like e.g. method groups and lambdas)

The option 2 didn't introduce any new classification. Values of fixed-size buffer types are classified as values of any other struct type.

### Obtaining instances of span types for a fixed-size buffer type

Since there is a guarantee that the first element in a fixed-size buffer type is aligned at the beginning of the type (no gap), compiler will use the
following code to get a ```Span``` value:
``` C#
MemoryMarshal.CreateSpan(ref Unsafe.As<TBuffer, TElement>(ref buffer), size)
```

And the following code to get a ```ReadOnlySpan``` value:
``` C#
MemoryMarshal.CreateReadOnlySpan(ref Unsafe.As<TBuffer, TElement>(ref Unsafe.AsRef(in buffer)), size)
```

In order to reduce IL size at use sites compiler should be able to add two generic reusable helpers into private implementation detail type and
use them across all use sites in the same program.

``` C#
public static System.Span<TElement> AsSpan<TBuffer, TElement>(ref TBuffer buffer, int size) where TBuffer : struct
{
    return MemoryMarshal.CreateSpan(ref Unsafe.As<TBuffer, TElement>(ref buffer), size);
}

public static System.ReadOnlySpan<TElement> AsReadOnlySpan<TBuffer, TElement>(in TBuffer buffer, int size) where TBuffer : struct
{
    return MemoryMarshal.CreateReadOnlySpan(ref Unsafe.As<TBuffer, TElement>(ref Unsafe.AsRef(in buffer)), size);
}
```

> Also, quite possibly compiler will be able to omit the ```Unsafe.As``` and ```Unsafe.AsRef``` calls in IL.

Author: > Once you do that, the `AsSpan` and `AsReadOnlySpan` helpers do not provide any IL size benefits.

After an offline discussion with @VSadov, I decided to not eliminate the calls. I will remove this statement from the spec.

> Also, quite possibly compiler will be able to omit the ```Unsafe.As``` and ```Unsafe.AsRef``` calls in IL.

Reviewer 3: Can you share the reasoning behind it? It looks be perfectly fine to me to skip these unnecessary `Unsafe.As` calls in the generated IL.

> Also, quite possibly compiler will be able to omit the ```Unsafe.As``` and ```Unsafe.AsRef``` calls in IL.

Reviewer 4: I'm curious about this as well. What was the rationale?

variable_declarators
    : variable_declarator (',' variable_declarator)*
    ;
    
variable_declarator
    : identifier ('=' variable_initializer)?
+   | fixed_size_buffer_declarator
    ;
    
fixed_size_buffer_declarator
    : identifier '[' constant_expression ']'
    ;    
```

A *fixed_size_buffer_declarator* introduces a new field and consists of an identifier that names it,
followed by a constant expression enclosed in `[` and `]` tokens. The type of the constant expression must be
implicitly convertible to type `int`, and the value must be a non-zero positive integer.

A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.

The field has a special struct type which represents a fixed-size buffer of a given element type.
The element type is the *type* specified in *field_declaration*.  The constant expression
denotes the number of elements in the buffer. The elements of a fixed-size buffer are laid out sequentially
in memory as though they are elements of an array.

The struct type will be defined using [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135) feature.

> The framework will likely include a set of predefined "buffer" types that cover a limited set of buffer sizes. When a predefined type doesn't exist,

Reviewer 3: What is the motivation for exposing predefined "buffer" types from the BCL? Is it just to save a bit of binary size or is it more than that?

Author: > What is the motivation for exposing predefined "buffer" types from the BCL? Is it just to save a bit of binary size or is it more than that?

The primary motivation is to avoid synthesizing types for common sizes in many user assemblies. There is also a benefit of interop between different user assemblies. If, for example, a buffer of size 10 type is in BCL, all user assemblies can easily pass instance of the same type around without a need to copy data between different versions of buffers of size 10.

> compiler will synthesize it in the module being built. Names of the generated types will be "speakable" in order to support consumption from other languages. 

Reviewer 4: I think we should reconsider this and not have them be speakable.  Related to Jan's feedback about System.Runtime.CompilerServices, I think we should instead:
- Have the compiler recognize whatever known public types we choose to expose from System.Runtime.CompilerServices
- Generate anything else it needs with internal unspeakable types.

This avoids conflicts with any types the user might declare, and avoids encouraging users to create types in System.Runtime.CompilerServices, which in general we don't want them doing.  I know and appreciate that one of the motivations for this was the libraries team creating those few built-in public helpers, but at most there will only be a handful of those and we can write them manually (even copy/pasting/tweaking what the compiler generated).

I also want the compiler to be free to add additional surface area to these types in the future.  For one of the same reason we don't publicly expose records, if the compiler is free to add additional APIs in the future, we wouldn't use it to create these public types, anyway.

If *primary_no_array_creation_expression* is a readonly variable, the result of evaluating a fixed-size buffer element access is a readonly variable
equaivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```[UnscopedRef] public readonly System.ReadOnlySpan<T> AsReadOnlySpan()```
method on *primary_no_array_creation_expression*. 

If *primary_no_array_creation_expression* is a value, the result of evaluating a fixed-size buffer element access is a value
equaivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```[UnscopedRef] public readonly System.ReadOnlySpan<T> AsReadOnlySpan()```
method on *primary_no_array_creation_expression*. 

For example:
``` C#

> void M1(Buffer10<int> x)

Reviewer 4: I personally find it strange that this magically-named type appears out of thin air for a field `public int buffer[10];`.  With fixed buffers today, the type of that field is effectively an int*... have we considered doing the same thing here and having the effective type be `Span<T>`?

Author: > I personally find it strange that this magically-named type appears out of thin air for a field `public int buffer[10];`.

As I mentioned on another thread, we don't have to use speakable names for types synthesized for field declarations. The purpose of the examples it to demonstrate the difference between a writable variable, a readonly variable, and a value. Note, that fields can be variables in these terms. That is a term that language specification uses for everything that has specific location in memory (locals, fields, by ref returns, array elements, etc.). The name of the type is not important and the examples can use fields without explicitly naming the types, but with the same exact effect. 

At the same time, one of the goals of this proposal is to support consumption of inline array types regardless of their names and regardless of where they locals, parameters, fields, return values, etc. Especially that anyone can declare types like that.

{

>     ref int a = ref x[0]; // Ok, equivalent to `ref int a = ref x.AsSpan()[0]`

Reviewer 4: The JIT should already be able to turn this into the equivalent of `ref x._element0`, but we should double-check it indeed does always do all the necessary inlining and the like.

}

void M2(in Buffer10<int> x)
{
    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
    System.Span<int> b = x; // An error, readonly mismatch
}

Buffer10<int> GetBuffer() => default;

void M3()
{
    System.ReadOnlySpan<int> a = GetBuffer(); // An error, ref-safety
    System.Span<int> b = GetBuffer(); // An error, ref-safety
}
```

With this approach, less context information is needed for conversion classification. This is going to significantly simplify implementation. 

### List patterns

[List patterns](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md) will not be supported for instances
of fixed-size buffer types.

### MemoryMarshal.CreateReadOnlySpan

We might want to specially recognize ```System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan``` method and allow
passing a reference to a readonly location as its first argument. This should allow users to define their own fixed-size buffer
types. For example, in order to be able to share a type across multiple assemblies and also include custom APIs into its definition.

> This might also help framework to define our shared fixed-size buffer types.

Reviewer 4: Developers can do this today with:
```C#
MemoryMarshal.CreateReadOnlySpan(ref Unsafe.AsRef(in location), length);
```
But this is also an example of an API that's motivation for allowing `ref readonly` on parameters; we would change CreateReadOnlySpan to accept a `ref readonly` instead of a `ref`, existing usage would continue to work, but it would also then work for creating a span around readonly fields.

  *primary_no_array_creation_expression* of an [element access](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#11710-element-access)
  with index of ```System.Index``` type, or of a type implicitly convertible to int. Result of the element access is a fixed-size member's element at the specified
  position, classified as a value.
- If access occurs in a context where direct assignments to an element of readonly fixed-size buffer are permitted, the result of the expression
  is classified as a value of type `System.Span<S>`, where S is the element type of the fixed-size buffer. The value can be used to access member's elements.
- Otherwise, the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of the fixed-size buffer.
  The value can be used to access member's elements.

### Definite assignment checking

Fixed-size buffers are not subject to definite assignment-checking, and fixed-size buffer members are ignored for purposes of definite-assignment checking of struct type variables.

> When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.

Reviewer 3: Does this is mean that the content will be uninitialized by default when the fixed-size buffer is a field of a local variable on stack? Is the program going to be able to see this uninitialized content without unsafe keyword?

Author: > Does this is mean that the content will be uninitialized by default when the fixed-size buffer is a field of a local variable on stack? Is the program going to be able to see this uninitialized content without unsafe keyword?

This proposal allows that, but we can require the buffers to be always initialized. Though, this is like to require additional changes in the language because this proposal doesn't support direct assignment to the safe fixed-size member in its entirety, and it might not be possible for compiler to verify that all elements in the member are definitely assigned (for example, when non-constant expressions are used to assign to elements).

    }

    [UnscopedRef]
    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
    {
        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
    }
}
```

The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
will be "speakable" in order to support consumption from other languages. 

A code generated for an access like: 
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer1[1] = val;
    }

    int M2()
    {
        return buffer2[1];
    }
}
```

will be equivalent to:
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer.AsSpan()[1] = val;
    }

    int M2()
    {
        return buffer2.AsReadOnlySpan()[1];
    }
}
```

#### Metadata import

> When compiler imports a field declaration of type *T* and the following conditions are all met:

Reviewer 3: If this going to kick even for types in the current compilation unit? If somebody manually field of the right shape manually, is it going to be treated as C# fixed size buffer?

Author: > Is this going to kick even for types in the current compilation unit? If somebody manually field of the right shape manually, is it going to be treated as C# fixed size buffer?

Yes. In general, we don't make distinction between types in source and  type in metadata for the purpose of deciding if they match a pattern.

- *T* is a struct type decorated with the ```InlineArray``` attribute, and
- The first instance field declared within *T* has type *F*, and
- There is a ```public System.Span<F> AsSpan()``` within *T*, and 

> - There is a ```public readonly System.ReadOnlySpan<T> AsReadOnlySpan()``` or ```public System.ReadOnlySpan<T> AsReadOnlySpan()``` within *T*. 

Reviewer 3: If the compiler is not going to validate that these AsSpan/AsReadOnlySpan methods have expected implementation, it opens possibility for some very creative hacks. For example, somebody can provide a custom AsReadOnlySpan implementation that returns variable sized Span. (Just an observation.)

Author: > If the compiler is not going to validate that these AsSpan/AsReadOnlySpan methods have expected implementation, it opens possibility for some very creative hacks. For example, somebody can provide a custom AsReadOnlySpan implementation that returns variable sized Span. (Just an observation.)

The validation of method bodies is not proposed. In general, compiler is not concerned about types violating a pattern they pretend to satisfy. There is an option for these helpers to be automatically provided by runtime on all inline array types with robust to hacks implementation. For example, array types have helpers like that to access element references, etc.

}

void M2(in Buffer10<int> x)
{
    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
    System.Span<int> b = x; // An error, readonly mismatch
}

Buffer10<int> GetBuffer() => default;

void M3()
{
    System.ReadOnlySpan<int> a = GetBuffer(); // An error, ref-safety
    System.Span<int> b = GetBuffer(); // An error, ref-safety
}
```

With this approach, less context information is needed for conversion classification. This is going to significantly simplify implementation. 

### List patterns

[List patterns](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md) will not be supported for instances
of fixed-size buffer types.

### Definite assignment checking

Regular definite assignment rules are applicable to variables that have a fixed-size buffer type. 

### An alternative to relying on AsSpan/AsReadOnlySpan helpers

If there is a guarantee that the first element in a fixed-size buffer type is aligned at the beginning of the type (no gap), then compiler can use the
following code to get a ```Span``` value (instead of relying on presence of ```AsSpan``` helper):
``` C#

> new Span<element type>(Unsafe.AsPointer(ref <the fixed-size buffer writable variable>), <size of the fixed-size buffer>)

Reviewer 3: Unsafe.AsPointer is only ok to use when the buffer is pinned. The buffer is not guaranteed to be pinned here.

This needs to use `MemoryMarshal.CreateSpan` API.

```

And the following code to get a ```ReadOnlySpan``` value (instead of relying on presence of ```AsReadOnlySpan``` helper):
``` C#
new ReadOnlySpan<element type>(Unsafe.AsPointer(ref Unsafe.AsRef(in <the fixed-size buffer writable variable>)), <size of the fixed-size buffer>)
```

> In order to reduce IL size at use sites compiler should be able to add two generic reusable helpers into private implementation detail type and

Reviewer 3: You can just directly emit `MemoryMarshal.CreateSpan` call inline and pass the address of the buffer and the length to it.

If you are emitting IL, you do not need the `Unsafe.AsPointer` or `Unsafe.AsRef` calls. These calls exist to make the code compile in C#. They are not necessary when emitting IL.

When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.

### Metadata

#### Metadata emit and code generation

For metadata encoding compiler will rely on recently added [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135). 

Fixed-size buffers like:
``` C#

> public struct C

Reviewer 6: Should this be `class C` to match the equivalent C# code below?

{
    public int buffer1[10];
    public readonly int buffer2[10];
}
```
will be emitted as  fields of a specially decorated struct type.

Equivalent C# code will be:

``` C#
public partial class C
{
    public Buffer10<int> buffer1;
    public readonly Buffer10<int> buffer2;
}

[System.Runtime.CompilerServices.InlineArray(10)]
public struct Buffer10<T>
{
    private T _element0;

    [UnscopedRef]
    public System.Span<T> AsSpan()
    {
        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
    }

    [UnscopedRef]
    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
    {
        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
    }
}
```

The actual naming conventions for the type and its members are TBD. Framework will likely to include a set of predefined "buffer" types that cover

> limited set of buffer sizes. When predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types

Reviewer 6: Perhaps:
> Framework will likely include a set of predefined "buffer" types that cover
> a limited set of buffer sizes. When a predefined type doesn't exist, ...

> will be "speakable" in order to support consumption from other languages. 

Reviewer 6: Will the synthesized type be generated as a top-level type with the same namespace and name as the expected predefined type? Will the synthesized type be `internal` rather than `public`?

        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
    }
}
```

The actual naming conventions for the type and its members are TBD. Framework will likely to include a set of predefined "buffer" types that cover
limited set of buffer sizes. When predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
will be "speakable" in order to support consumption from other languages. 

A code generated for an access like: 
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer1[1] = val;
    }

    int M2()
    {
        return buffer2[1];
    }
}
```

will be equivalent to:
``` C#
public partial class C
{
    void M1(int val)
    {
        buffer.AsSpan()[1] = val;
    }

    int M2()
    {
        return buffer2.AsReadOnlySpan()[1];
    }
}
```

#### Metadata import

When compiler imports a field declaration of type *T* and the following conditions are all met:
- *T* is a struct type decorated with the ```InlineArray``` attribute, and

> - The first instance field declared within *T* has type *F*, and

Reviewer 5: Shouldn't there be exactly one field in *T*?
